test/Test/Pos/Block/Logic/VarSpec.hs

```diff
@@ -283,3 +296,63 @@ verifyBlockScenarioResult = \case
             "Block scenario resulted in a change to the blockchain" <>
             " relative to the " <> show snapId <> " snapshot:\n" <>
             show dbDiff
+
+----------------------------------------------------------------------------
+-- Forks
+----------------------------------------------------------------------------
+
+singleForkSpec :: ForkDepth -> Spec
+singleForkSpec fd = do
+    prop singleForkDesc (singleForkProp fd)
+  where
+    singleForkDesc =
+      "a blockchain of length q=0..(9.5*k) blocks can switch to a fork " <>
+      "of length j>i with a common prefix i, d=q-i"
+
+singleForkProp :: ForkDepth -> BlockProperty ()
+singleForkProp fd = do
+    scenario <- blockPropertyScenarioGen $ genSingleFork fd
+    runBlockScenarioAndVerify scenario
+
+data ForkDepth = ForkShort | ForkMedium | ForkDeep
+
+genSingleFork :: forall g m. (RandomGen g, Monad m) => ForkDepth -> BlockEventGenT g m ()
+genSingleFork fd = do
+    let k = fromIntegral blkSecurityParam
+    d <- getRandomR $ case fd of
+        ForkShort  -> (1, k - 1)
+        ForkMedium -> (max 1 (k-2), k+2)
+        ForkDeep   -> (k+1, div (k*3) 2)
+    -- the depth must be <=k for a successful rollback.
+    let expectSuccess = d <= k
+    -- original blockchain max index q<(9.5*k)
+    q <- getRandomR (d+1, 9 * k + div k 2)
+    let
+        -- max index of the common prefix. i>0 because d<q
+        i = q-d
+    -- fork blockchain max index j>i. the upper bound is arbitrary.
+    -- dj=j-i
+    dj <- getRandomR (1, d*2)
+    -- now we can generate paths
+    --
+    -- B0 - B1 - B2 - B3 - B4 - B5 - B6 - B7
+    --              \
+    --                C3 - C4 - C5 - C6
```

Comment:
By this moment it's hard to keep track of what is `d`, `dj`, `i`, `q`, etc. If you somehow illustrate them with this picture (at least write their values for this case), it will make it easier to understand what's going on.